File: instrumentation/android-log/agent/src/main/java/io/opentelemetry/instrumentation/agent/log/AndroidLogPlugin.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.agent.log;

import static net.bytebuddy.matcher.ElementMatchers.is;
import static net.bytebuddy.matcher.ElementMatchers.isMethod;

import android.util.Log;
import io.opentelemetry.instrumentation.library.log.AndroidLogSubstitutions;
import java.io.IOException;
import net.bytebuddy.asm.MemberSubstitution;
import net.bytebuddy.build.AndroidDescriptor;
import net.bytebuddy.build.Plugin;
import net.bytebuddy.description.type.TypeDescription;
import net.bytebuddy.dynamic.ClassFileLocator;
import net.bytebuddy.dynamic.DynamicType;

public class AndroidLogPlugin implements Plugin {
    private final AndroidDescriptor androidDescriptor;

    public AndroidLogPlugin(AndroidDescriptor androidDescriptor) {
        this.androidDescriptor = androidDescriptor;
    }

    @Override
    public DynamicType.Builder<?> apply(
            DynamicType.Builder<?> builder,
            TypeDescription typeDescription,
            ClassFileLocator classFileLocator) {
        try {
            return builder.visit(
                    MemberSubstitution.relaxed()
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "v", String.class, String.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForVerbose", String.class, String.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "v",
                                                    String.class,
                                                    String.class,
                                                    Throwable.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForVerbose2",
                                            String.class,
                                            String.class,
                                            Throwable.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "d", String.class, String.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForDebug", String.class, String.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "d",
                                                    String.class,
                                                    String.class,
                                                    Throwable.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForDebug2",
                                            String.class,
                                            String.class,
                                            Throwable.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "i", String.class, String.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForInfo", String.class, String.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "i",
                                                    String.class,
                                                    String.class,
                                                    Throwable.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForInfo2",
                                            String.class,
                                            String.class,
                                            Throwable.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "w", String.class, String.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForWarn", String.class, String.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "w", String.class, Throwable.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForWarn2", String.class, Throwable.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "w",
                                                    String.class,
                                                    String.class,
                                                    Throwable.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForWarn3",
                                            String.class,
                                            String.class,
                                            Throwable.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "e", String.class, String.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForError", String.class, String.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "e",
                                                    String.class,
                                                    String.class,
                                                    Throwable.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForError2",
                                            String.class,
                                            String.class,
                                            Throwable.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "wtf", String.class, String.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForWtf", String.class, String.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "wtf", String.class, Throwable.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForWtf2", String.class, Throwable.class))
                            .method(
                                    is(
                                            Log.class.getDeclaredMethod(
                                                    "wtf",
                                                    String.class,
                                                    String.class,
                                                    Throwable.class)))
                            .replaceWith(
                                    AndroidLogSubstitutions.class.getDeclaredMethod(
                                            "substitutionForWtf3",
                                            String.class,
                                            String.class,
                                            Throwable.class))
                            .on(isMethod()));

        } catch (NoSuchMethodException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void close() throws IOException {
        // No operation.
    }

    @Override
    public boolean matches(TypeDescription target) {
        return androidDescriptor.getTypeScope(target) == AndroidDescriptor.TypeScope.LOCAL;
    }
}

File: instrumentation/android-log/agent/src/main/java/io/opentelemetry/instrumentation/agent/log/AndroidLogPlugin.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.instrumentation.agent.log

import android.util.Log
import io.opentelemetry.instrumentation.library.log.AndroidLogSubstitutions
import net.bytebuddy.asm.MemberSubstitution
import net.bytebuddy.build.AndroidDescriptor
import net.bytebuddy.build.Plugin
import net.bytebuddy.description.type.TypeDescription
import net.bytebuddy.dynamic.ClassFileLocator
import net.bytebuddy.dynamic.DynamicType
import net.bytebuddy.matcher.ElementMatchers
import java.io.IOException

internal class AndroidLogPlugin(
    private val androidDescriptor: AndroidDescriptor,
) : Plugin {
    override fun apply(
        builder: DynamicType.Builder<*>,
        typeDescription: TypeDescription,
        classFileLocator: ClassFileLocator,
    ): DynamicType.Builder<*> {
        try {
            return builder.visit(
                MemberSubstitution
                    .relaxed()
                    .method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "v",
                                String::class.java,
                                String::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForVerbose",
                            String::class.java,
                            String::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "v",
                                String::class.java,
                                String::class.java,
                                Throwable::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForVerbose2",
                            String::class.java,
                            String::class.java,
                            Throwable::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "d",
                                String::class.java,
                                String::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForDebug",
                            String::class.java,
                            String::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "d",
                                String::class.java,
                                String::class.java,
                                Throwable::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForDebug2",
                            String::class.java,
                            String::class.java,
                            Throwable::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "i",
                                String::class.java,
                                String::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForInfo",
                            String::class.java,
                            String::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "i",
                                String::class.java,
                                String::class.java,
                                Throwable::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForInfo2",
                            String::class.java,
                            String::class.java,
                            Throwable::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "w",
                                String::class.java,
                                String::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForWarn",
                            String::class.java,
                            String::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "w",
                                String::class.java,
                                Throwable::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForWarn2",
                            String::class.java,
                            Throwable::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "w",
                                String::class.java,
                                String::class.java,
                                Throwable::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForWarn3",
                            String::class.java,
                            String::class.java,
                            Throwable::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "e",
                                String::class.java,
                                String::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForError",
                            String::class.java,
                            String::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "e",
                                String::class.java,
                                String::class.java,
                                Throwable::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForError2",
                            String::class.java,
                            String::class.java,
                            Throwable::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "wtf",
                                String::class.java,
                                String::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForWtf",
                            String::class.java,
                            String::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "wtf",
                                String::class.java,
                                Throwable::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForWtf2",
                            String::class.java,
                            Throwable::class.java,
                        ),
                    ).method(
                        ElementMatchers.`is`(
                            Log::class.java.getDeclaredMethod(
                                "wtf",
                                String::class.java,
                                String::class.java,
                                Throwable::class.java,
                            ),
                        ),
                    ).replaceWith(
                        AndroidLogSubstitutions::class.java.getDeclaredMethod(
                            "substitutionForWtf3",
                            String::class.java,
                            String::class.java,
                            Throwable::class.java,
                        ),
                    ).on(ElementMatchers.isMethod()),
            )
        } catch (e: NoSuchMethodException) {
            throw RuntimeException(e)
        }
    }

    @Throws(IOException::class)
    override fun close() {
        // No operation.
    }

    override fun matches(target: TypeDescription): Boolean = androidDescriptor.getTypeScope(target) == AndroidDescriptor.TypeScope.LOCAL
}
